cache_system/src/backend/policy/remove_if.rs

```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+impl<K, V> Subscriber for RemoveIfPolicy<K, V>
+where
+    K: Clone + Eq + Debug + Hash + Ord + Send + 'static,
+    V: Clone + Debug + Send + 'static,
+{
+    type K = K;
+    type V = V;
+}
+
+/// Handle created by [`RemoveIfPolicy`] that can be used to evict data from caches.
+#[derive(Debug, Clone)]
+pub struct RemoveIfHandle<K, V>
+where
+    K: Clone + Eq + Debug + Hash + Ord + Send + 'static,
+    V: Clone + Debug + Send + 'static,
+{
+    callback_handle: Arc<Mutex<Option<CallbackHandle<K, V>>>>,
+    metric_removed_by_predicate: U64Counter,
+}
+
+impl<K, V> RemoveIfHandle<K, V>
+where
+    K: Clone + Eq + Debug + Hash + Ord + Send + 'static,
+    V: Clone + Debug + Send + 'static,
+{
+    /// "remove" a key (aka remove it from the shared backend) if the
+    /// specified predicate is true. If the key is removed return
+    /// true, otherwise return false
+    ///
+    /// Note that the predicate function is called while the lock is
+    /// held (and thus the inner backend can't be concurrently accessed
+    pub fn remove_if<P>(&self, k: &K, predicate: P) -> bool
+    where
+        P: Fn(V) -> bool + Send,
+    {
+        let mut guard = self.callback_handle.lock();
+        let handle = match guard.as_mut() {
+            Some(handle) => handle,
+            None => return false,
+        };
+
+        let metric_removed_by_predicate = self.metric_removed_by_predicate.clone();
+        let mut removed = false;
+        let removed_captured = &mut removed;
+        let k = k.clone();
+        handle.init_requests(vec![ChangeRequest::from_fn(move |backend| {
+            if let Some(v) = backend.get(&k) {
+                if predicate(v) {
+                    metric_removed_by_predicate.inc(1);
+                    backend.remove(&k);
+                    *removed_captured = true;
+                }
+            }
+        })]);
+
+        removed
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use metric::{Observation, RawReporter};
+
+    use crate::backend::{policy::PolicyBackend, CacheBackend};
+
+    use super::*;
+
+    #[test]
+    fn test_generic_backend() {
+        use crate::backend::test_util::test_generic;
+
+        test_generic(|| {
+            let metric_registry = metric::Registry::new();
+            let mut backend = PolicyBackend::new(Box::new(HashMap::<u8, String>::new()));
+            let (constructor, _handle) =
+                RemoveIfPolicy::create_constructor_and_handle("my_cache", &metric_registry);
+            backend.add_policy(constructor);
+            backend
+        });
+    }
+
+    #[test]
+    fn test_remove_if() {
+        let metric_registry = metric::Registry::new();
+        let mut backend = PolicyBackend::new(Box::new(HashMap::<u8, String>::new()));
+        let (constructor, handle) =
+            RemoveIfPolicy::create_constructor_and_handle("my_cache", &metric_registry);
+        backend.add_policy(constructor);
```

Comment:
```suggestion
        let (policy, handle) =
            RemoveIfPolicy::create_constructor_and_handle("my_cache", &metric_registry);
        backend.add_policy(policy);
```

Which I realize isn't quite right as the `constructor` isn't a policy but a `Fn` object. But then at least it matches the name `add_policy`  -- I was just thinking that when reading this code the difference between constructor and policy likely isn't relevant to understanding how to use the code. Just a thought I had 🤷

Reply:
Went for `policy_constructor` which I think is a good middle-ground between "what it is" and "easy to understand/read".